File: sort_algorithm/heap_sort.go
package sort_algorithm

import "fmt"

type HeapSort []int

func (h HeapSort) SortImpl() {
	n := len(h)
	for i := n/2 - 1; i >= 0; i-- {
		h.heapify(h, i, n)
	}
	for i := n - 1; i > 0; i-- {
		h[i], h[0] = h[0], h[i]
		h.heapify(h, 0, i)
	}
	fmt.Println(h)
}

func (h HeapSort) heapify(arr HeapSort, l, n int) {
	tmp := arr[l]
	for i := 2*l + 1; i < n; i = 2*l + 1 {
		if i+1 < n && arr[i] < arr[i+1] {
			i++
		}
		if arr[i] <= tmp {
			break
		}
		arr[l] = arr[i]
		l = i
	}

	arr[l] = tmp
}

File: java/sort_algorithm/impl/HeapSort.java
package sort_algorithm.impl;

import sort_algorithm.Sort;

public class HeapSort implements Sort {
    @Override
    public void sort(int[] arr) {
        for (int i = arr.length / 2 - 1; i >= 0; i--) {
            heapify(arr, i, arr.length);
        }
        for (int i = arr.length - 1; i > 0 ; --i) {
            int tmp = arr[0];
            arr[0] = arr[i];
            arr[i] = tmp;
            heapify(arr, 0, arr.length - i);
        }
    }
    public void heapify(int[] arr, int l, int r) {
        int val = arr[l];
        for (int i = 2 * l + 1; i < r; i = 2 * l + 1) {
            if (i + 1 < r && arr[i+1] > arr[i]) i++;
            if (arr[i] <= val) break;
            arr[l] = arr[i];
            l = i;
        }
        arr[l] = val;
    }
}
